grammar.js

```diff
@@ -190,10 +190,12 @@ module.exports = grammar({
       '$', '(', repeat($._tokens), ')', optional(/[^+*?]+/), choice('+', '*', '?')
     ),
 
+    non_special_punctuation: $ => /[/_\-=->,;:::!=?.@*&#%^+<>|~]+/,
+
     _non_special_token: $ => choice(
       $._literal, $.identifier, $.metavariable, $.mutable_specifier, $.self, $.super, $.crate,
       alias(choice(...primitive_types), $.primitive_type),
-      /[/_\-=->,;:::!=?.@*&#%^+<>|~]+/,
+      $.non_special_punctuation,
```

Comment:
I think that if we're going to make this visible, it would be better to parse these tokens the same as when they appear in other contexts. I believe that in Rust macros, the token trees need to consist of valid Rust tokens. So we don't need a separate, catch-all regex for handling all of these. Instead, we should just list out all of the valid punctuation tokens. They can be found in [the language reference](https://doc.rust-lang.org/reference/tokens.html#punctuation).

```js
{
  _non_special_token: $ => choice(
    // named tokens...
    // keywords...
    '+', '-', '*', '/', '%', '^', '!',  // all of the other punctuation tokens...
  )
}
```

This will actually simplify the parser, since we'll be reusing existing tokens and can get rid of this catch-all regex completely.

Reply:
Once you do this, the corpus files will require much smaller changes, since these tokens will be anonymous nodes, just like they are in other contexts.